File: 简单/剑指 Offer 30. 包含min函数的栈.py
"""

定义栈的数据结构，请在该类型中实现一个能够得到栈的最小元素的 min 函数在该栈中，
调用 min、push 及 pop 的时间复杂度都是 O(1)。


示例:
MinStack minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
minStack.min();   --> 返回 -3.
minStack.pop();
minStack.top();      --> 返回 0.
minStack.min();   --> 返回 -2.

"""


class MinStack:

    def __init__(self):
        self.stack = []
        self.auxiliary_stack = []

    def push(self, x):
        if not self.stack:
            self.stack.append(x)
            self.auxiliary_stack.append(x)
        else:
            self.stack.append(x)
            if x <= self.auxiliary_stack[-1]:
                self.auxiliary_stack.append(x)

    def pop(self):
        if self.auxiliary_stack[-1] == self.stack[-1]:
            self.auxiliary_stack.pop()
        self.stack.pop()

    def top(self):
        return self.stack[-1]

    def min(self):
        return self.auxiliary_stack[-1]

File: 剑指offer/day1-栈/剑指 Offer 30. 包含min函数的栈.js
// 定义栈的数据结构，请在该类型中实现一个能够得到栈的最小元素的 min 函数在该栈中，调用 min、push 及 pop 的时间复杂度都是 O(1)。

// 示例:
// MinStack minStack = new MinStack();
// minStack.push(-2);
// minStack.push(0);
// minStack.push(-3);
// minStack.min(); --> 返回 - 3.
// minStack.pop();
// minStack.top(); --> 返回 0.
// minStack.min(); --> 返回 - 2.


/**
 * initialize your data structure here.
 */
var MinStack = function () {
  this.stack = []
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function (x) {
  this.stack.push(x)
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
  if (this.stack.length > 0) {
    this.stack.pop()
  }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  return this.stack[this.stack.length - 1]
};

/**
 * @return {number}
 */
MinStack.prototype.min = function () {
  return Math.min(...this.stack)
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.min()
 */
